prompb/types.proto

```diff
@@ -54,13 +54,58 @@ message Exemplar {
   int64 timestamp = 3;
 }
 
+// A sparse histogram.
+message Histogram {
+  enum ResetHint {
+    UNKNOWN = 0; // Need to test for a counter reset explicitly.
+    YES     = 1; // This is the 1st histogram after a counter reset.
+    NO      = 2; // There was no counter reset between this and the previous Histogram.
+    GAUGE   = 3; // This is a gauge histograms where counter resets don't happen.
+  }
+  
+  uint64 count       = 1;
+  double count_float = 2; // Overrides count if > 0.
+  double sum         = 3;
+  // The schema defines the bucket schema. Currently, valid numbers
+  // are -4 <= n <= 8. They are all for base-2 bucket schemas, where 1
+  // is a bucket boundary in each case, and then each power of two is
+  // divided into 2^n logarithmic buckets. Or in other words, each
+  // bucket boundary is the previous boundary times 2^(2^-n). In the
+  // future, more bucket schemas may be added using numbers < -4 or >
+  // 8.
+  sint32 schema             = 4;
+  double zero_threshold     = 5; // Breadth of the zero bucket.
+  uint64 zero_count         = 6; // Count in zero bucket.
+  double zero_count_float   = 7; // Overrides sb_zero_count if > 0.
+  Buckets negative_buckets  = 8;
+  Buckets positive_buckets  = 9;
+  ResetHint reset_hint      = 10;
+  // timestamp is in ms format, see model/timestamp/timestamp.go for
+  // conversion from time.Time to Prometheus timestamp.
+  int64 timestamp = 11;
+} 
+
+// Sparse buckets.
+message Buckets {
+  message Span {
+    sint32 offset = 1; // Gap to previous span, or starting point for 1st span (which can be negative).
+    uint32 length = 2; // Length of consecutive buckets.
+  }
+  repeated Span span = 1;
+  // Only one of "delta" or "count" may be used, the former for regular
+  // histograms with integer counts, the latter for float histograms.
+  repeated sint64 delta = 2; // Count delta of each bucket compared to previous one (or to zero for 1st bucket).
```

Comment:
If I understand correctly, logically these values belong with the `Span`, but as an optimisation to reduce Protobuf metadata (tag and length) you have pulled them up one level and concatenated them, so the encoder/decoder must do a little extra work.
Either way, I think it would be worth some more comments so readers can understand.

Reply:
I think you are quite right. I'll add more comments.

For the decoder, it's actually less work, because the protobuf layout happens to be the same as the layout in TSDB. (Which isn't a reason to do it that way on its own, of course, just a convenient coincidence.)

Reply:
For remote write, it's actually the TSDB layout on both sides, so it's even less work for the encoder… (Encoding is more costly for exposition, depending on the way the instrumentation library manages the buckets internally.)